From dc80fb9abf00c7818a062f0f65e609aa5e198e0d Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:47 -0400
Subject: [virt] kvm/nvmx: Update guest activity state field on L2 exits

Message-id: <1406040016-3289-133-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86211
O-Subject: [RHEL7 PATCH v2 132/221] KVM: nVMX: Update guest activity state field on L2 exits
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

Set guest activity state in L1's VMCS according to the VCPUs mp_state.
This ensures we report the correct state in case we L2 executed HLT or
if we put L2 into HLT state and it was now woken up by an event.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 3edf1e698ff638c2ab095e8c60fd11c5d292fc5f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 294ab66..c9ceb5e 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8219,6 +8219,10 @@ static void prepare_vmcs12(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12,
   vmcs_read32(GUEST_INTERRUPTIBILITY_INFO);
  vmcs12->guest_pending_dbg_exceptions =
   vmcs_readl(GUEST_PENDING_DBG_EXCEPTIONS);
+ if (vcpu->arch.mp_state == KVM_MP_STATE_HALTED)
+  vmcs12->guest_activity_state = GUEST_ACTIVITY_HLT;
+ else
+  vmcs12->guest_activity_state = GUEST_ACTIVITY_ACTIVE;
 
  if ((vmcs12->pin_based_vm_exec_control & PIN_BASED_VMX_PREEMPTION_TIMER) &&
      (vmcs12->vm_exit_controls & VM_EXIT_SAVE_VMX_PREEMPTION_TIMER))
-- 
1.7.1